apps/src/templates/instructions/TeacherFeedbackKeepWorking.jsx

```diff
@@ -0,0 +1,127 @@
+import React, {Component} from 'react';
+import PropTypes from 'prop-types';
+import i18n from '@cdo/locale';
+import color from '@cdo/apps/util/color';
+import ReactTooltip from 'react-tooltip';
+import {makeEnum} from '@cdo/apps/utils';
+
+const ReviewStates = makeEnum('completed', 'keepWorking');
+
+// TeacherFeedbackKeepWorking displays a checkbox which can be in one of 3 states:
+// 1. Checked - meaning the teacher has requested the student to keep working
+// 2. Unchecked - meaning the teacher has not requested the student to keep working, or has removed the previous request
+// 3. Indeterminate - meaning the level is awaiting teacher review (the teacher requested
+// the student to keep working and the student has made progress since that feedback was given)
+// This checkbox is displayed to and controlled by the teacher.
+class TeacherFeedbackKeepWorking extends Component {
+  static propTypes = {
+    latestFeedback: PropTypes.object,
+    reviewState: PropTypes.oneOf(Object.values(ReviewStates)),
+    setReviewState: PropTypes.func,
+    setReviewStateChanged: PropTypes.func
+  };
+
+  checkbox = null;
+  isAwaitingTeacherReview =
+    this.props.latestFeedback.review_state === ReviewStates.keepWorking &&
+    this.props.latestFeedback.student_updated_since_feedback;
+
+  componentDidMount() {
+    if (this.isAwaitingTeacherReview) {
+      this.checkbox.indeterminate = true;
+    }
+  }
+
+  onCheckboxChange = () => {
+    const newReviewState = this.getNewReviewState();
+    this.props.setReviewState(newReviewState);
+
+    const isDifferentFromInitial = this.isDifferentFromInitial(newReviewState);
+    this.props.setReviewStateChanged(isDifferentFromInitial);
+  };
+
+  getNewReviewState() {
```

Comment:
Mostly curious, did you consider basing this logic on reviewState instead of the the checkbox state?

I think one of the things that I find a little bit confusing in this component is the mapping back and forth between the three conceptual states ('keep working', 'completed', 'awaiting review') and the three checkbox states ('checked', 'unchecked', 'indeterminate').  There is a 1:1 mapping between these, right?  If so, it might be nice to work primarily in one of these (the conceptual state?) and do the mapping only in display logic.

Reply:
It's not exactly a 1:1 because unchecked = completed or null depending on previous states

Reply:
I've tried to refactor this several times and couldn't come up with something that was more straight forward. If you'd like to pair on a refactor, I'm super open to that!!

Reply:
Ah, this is the key that I was missing -- there are *four* conceptual review states ('no value', 'keep working', 'completed', and 'awaiting review') and three checkbox states.  Thinking...